source/daplink/drag-n-drop/virtual_fs.c

```diff
@@ -295,8 +295,19 @@ void vfs_init(const vfs_filename_t drive_name, uint32_t disk_size)
     init_complete = false;
     // Initialize MBR
     memcpy(&mbr, &mbr_tmpl, sizeof(mbr_t));
-    mbr.total_logical_sectors = ((disk_size + KB(64)) / mbr.bytes_per_sector);
-    mbr.logical_sectors_per_fat = (3 * (((mbr.total_logical_sectors / mbr.sectors_per_cluster) + 1023) / 1024));
+    if (((disk_size + KB(64)) / mbr.bytes_per_sector) >= 32768) {
+        mbr.total_logical_sectors = 0;
+        mbr.big_sectors_on_drive = ((disk_size + KB(64)) / mbr.bytes_per_sector);
+        util_assert((mbr.big_sectors_on_drive / 4096) <= 64);
+        mbr.sectors_per_cluster = 0x10;
+        while ((mbr.big_sectors_on_drive / mbr.sectors_per_cluster) >= 4096) {
+            mbr.sectors_per_cluster *= 2;
+        }
+        mbr.logical_sectors_per_fat = (3 * (((mbr.big_sectors_on_drive / mbr.sectors_per_cluster) + 1023) / 1024));
+    } else {
+        mbr.total_logical_sectors = ((disk_size + KB(64)) / mbr.bytes_per_sector);
+        mbr.logical_sectors_per_fat = (3 * (((mbr.total_logical_sectors / mbr.sectors_per_cluster) + 1023) / 1024));
+    }
```

Comment:
The goal was not to change to FAT16, merely to get support for a large file system and with the existing setup the number of clusters became too high. I looked at the original code in  https://github.com/embeddedartists/CMSIS-DAP/blob/master/interface/Common/src/usbd_user_msc.c which does this conversion. The boards I have added only actually needs ca 17Mb file systems but the comment in vfs_user.c said to use 4x max flash which gave me a total size of 68Mb. I don't know what the extra 51Mb are for.

I think a proper change to FAT16 could be beneficial as the file system requirements and flash sizes will always increase.